File: 1962.remove-stones-to-minimize-the-total.py
#
# @lc app=leetcode id=1962 lang=python3
#
# [1962] Remove Stones to Minimize the Total
#

# @lc code=start
import heapq
from typing import List


class Solution:
    # clearly a priority queue question
    def minStoneSum(self, piles: List[int], k: int) -> int:
        heap = []
        for pile in piles:
            heapq.heappush(heap, -pile)
        for i in range(k):
            item = - heapq.heappop(heap)
            item -= item // 2
            heapq.heappush(heap, -item)
        return - sum(heap)
# @lc code=end

File: 1962.remove-stones-to-minimize-the-total.cs
/*
 * @lc app=leetcode id=1962 lang=csharp
 *
 * [1962] Remove Stones to Minimize the Total
 */

// @lc code=start
public class Solution 
{
    public int MinStoneSum(int[] piles, int k) 
    {
        PriorityQueue<int, int> heap = new PriorityQueue<int, int>();
        foreach (int pile in piles)
        {
            heap.Enqueue(pile, - pile);

        }

        for (int i = 0; i < k; i++)
        {
            int item = heap.Dequeue();
            item = item - item / 2;
            heap.Enqueue(item, - item);
        }

        int result = 0;
        while (heap.TryDequeue(out int pile, out int priority))
        {
            result += pile;
        }
        return result;
    }
}
// @lc code=end

